File: 2048.py
import random
from abc import ABC, abstractmethod

# ==================== BLOCK (TILE) CLASS ====================

class Block:
    """Represents a single tile in the 2048 game."""
    
    def __init__(self, x: int, y: int, value: int = 2):
        self.value = value
        self.position = (x, y)

    def update_value(self):
        """Double the value when merged."""
        self.value *= 2


# ==================== BLOCK FACTORY ====================

class BlockFactory:
    """Factory to create new blocks dynamically."""

    @staticmethod
    def create_block(x, y):
        return Block(x, y, random.choice([2, 4]))  # New blocks are usually 2 or 4


# ==================== MOVEMENT STRATEGY INTERFACE ====================

class MoveStrategy(ABC):
    """Abstract Strategy for movement directions."""

    @abstractmethod
    def move(self, grid):
        pass


# ==================== MOVEMENT STRATEGY IMPLEMENTATIONS ====================

class MoveUp(MoveStrategy):
    """Handles movement in the UP direction."""

    def move(self, grid):
        return grid.shift_and_merge(direction="up")


class MoveDown(MoveStrategy):
    """Handles movement in the DOWN direction."""

    def move(self, grid):
        return grid.shift_and_merge(direction="down")


class MoveLeft(MoveStrategy):
    """Handles movement in the LEFT direction."""

    def move(self, grid):
        return grid.shift_and_merge(direction="left")


class MoveRight(MoveStrategy):
    """Handles movement in the RIGHT direction."""

    def move(self, grid):
        return grid.shift_and_merge(direction="right")


# ==================== GRID CLASS ====================

class Grid:
    """Represents the 2048 game board."""
    
    def __init__(self, size: int = 4):
        self.size = size
        self.grid = [[None for _ in range(size)] for _ in range(size)]
        self.spawn_new_tile()
        self.spawn_new_tile()

    def spawn_new_tile(self):
        """Places a new tile in a random empty position."""
        empty_positions = [(i, j) for i in range(self.size) for j in range(self.size) if self.grid[i][j] is None]
        if empty_positions:
            x, y = random.choice(empty_positions)
            self.grid[x][y] = BlockFactory.create_block(x, y)

    def is_valid_move(self, direction):
        """Check if a move is possible in the given direction."""
        for i in range(self.size):
            for j in range(self.size):
                if self.grid[i][j] is None:
                    continue
                new_i, new_j = self.get_new_position(i, j, direction)
                if 0 <= new_i < self.size and 0 <= new_j < self.size:
                    if self.grid[new_i][new_j] is None or self.grid[new_i][new_j].value == self.grid[i][j].value:
                        return True
        return False

    def get_new_position(self, i, j, direction):
        """Get the new position after movement."""
        if direction == "up":
            return i - 1, j
        elif direction == "down":
            return i + 1, j
        elif direction == "left":
            return i, j - 1
        elif direction == "right":
            return i, j + 1
        return i, j

    def shift_and_merge(self, direction):
        """Shift and merge tiles in the given direction."""
        if not self.is_valid_move(direction):
            return False  # No movement possible

        merged = [[False] * self.size for _ in range(self.size)]
        range_order = range(self.size) if direction in ["down", "right"] else range(self.size - 1, -1, -1)

        for i in range_order:
            for j in range_order:
                if self.grid[i][j] is None:
                    continue
                ni, nj = self.get_new_position(i, j, direction)
                
                while 0 <= ni < self.size and 0 <= nj < self.size and self.grid[ni][nj] is None:
                    self.grid[ni][nj], self.grid[i][j] = self.grid[i][j], None
                    i, j = ni, nj
                    ni, nj = self.get_new_position(i, j, direction)

                if (0 <= ni < self.size and 0 <= nj < self.size and 
                    self.grid[ni][nj] is not None and 
                    self.grid[ni][nj].value == self.grid[i][j].value and 
                    not merged[ni][nj]):
                    
                    self.grid[ni][nj].update_value()
                    self.grid[i][j] = None
                    merged[ni][nj] = True

        self.spawn_new_tile()
        return True

    def print_grid(self):
        """Print the grid for debugging."""
        for row in self.grid:
            print([block.value if block else 0 for block in row])
        print("\n")


# ==================== GAME CONTROLLER ====================

class GameController:
    """Manages game state and player moves."""
    
    def __init__(self):
        self.grid = Grid()
        self.moves = {
            "up": MoveUp(),
            "down": MoveDown(),
            "left": MoveLeft(),
            "right": MoveRight()
        }

    def play_move(self, direction):
        """Process player move."""
        if direction not in self.moves:
            print("Invalid move")
            return

        if self.moves[direction].move(self.grid):
            self.grid.print_grid()
        else:
            print("Move not possible!")


# ==================== MAIN GAME LOOP ====================

if __name__ == "__main__":
    game = GameController()
    
    while True:
        move = input("Enter move (up/down/left/right): ").strip().lower()
        game.play_move(move)

File: 2048.java
import java.util.*;


// ==================== BLOCK CLASS ====================
class Block {
    private int value;
    private int x, y;

    public Block(int x, int y, int value) {
        this.x = x;
        this.y = y;
        this.value = value;
    }

    public int getValue() {
        return value;
    }

    public void doubleValue() {
        this.value *= 2;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public void setPosition(int x, int y) {
        this.x = x;
        this.y = y;
    }
}


// ==================== BLOCK FACTORY ====================
class BlockFactory {
    private static final Random random = new Random();

    public static Block createBlock(int x, int y) {
        return new Block(x, y, random.nextInt(10) < 9 ? 2 : 4);
    }
}


// ==================== MOVEMENT STRATEGY INTERFACE ====================
interface MoveStrategy {
    boolean move(Grid grid);
}


// ==================== MOVEMENT STRATEGY IMPLEMENTATIONS ====================
class MoveUp implements MoveStrategy {
    @Override
    public boolean move(Grid grid) {
        return grid.shiftAndMerge("up");
    }
}

class MoveDown implements MoveStrategy {
    @Override
    public boolean move(Grid grid) {
        return grid.shiftAndMerge("down");
    }
}

class MoveLeft implements MoveStrategy {
    @Override
    public boolean move(Grid grid) {
        return grid.shiftAndMerge("left");
    }
}

class MoveRight implements MoveStrategy {
    @Override
    public boolean move(Grid grid) {
        return grid.shiftAndMerge("right");
    }
}


// ==================== GRID CLASS ====================
class Grid {
    private final int size;
    private final Block[][] grid;
    private final Random random = new Random();

    public Grid(int size) {
        this.size = size;
        this.grid = new Block[size][size];
        spawnNewTile();
        spawnNewTile();
    }

    public void spawnNewTile() {
        List<int[]> emptyPositions = new ArrayList<>();
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                if (grid[i][j] == null) {
                    emptyPositions.add(new int[]{i, j});
                }
            }
        }

        if (!emptyPositions.isEmpty()) {
            int[] pos = emptyPositions.get(random.nextInt(emptyPositions.size()));
            grid[pos[0]][pos[1]] = BlockFactory.createBlock(pos[0], pos[1]);
        }
    }

    public boolean isValidMove(String direction) {
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                if (grid[i][j] == null) continue;

                int[] newPos = getNewPosition(i, j, direction);
                int ni = newPos[0], nj = newPos[1];

                if (ni >= 0 && ni < size && nj >= 0 && nj < size) {
                    if (grid[ni][nj] == null || grid[ni][nj].getValue() == grid[i][j].getValue()) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    private int[] getNewPosition(int i, int j, String direction) {
        switch (direction) {
            case "up": return new int[]{i - 1, j};
            case "down": return new int[]{i + 1, j};
            case "left": return new int[]{i, j - 1};
            case "right": return new int[]{i, j + 1};
            default: return new int[]{i, j};
        }
    }

    public boolean shiftAndMerge(String direction) {
        if (!isValidMove(direction)) return false;

        boolean[][] merged = new boolean[size][size];
        int start = (direction.equals("down") || direction.equals("right")) ? size - 1 : 0;
        int step = (direction.equals("down") || direction.equals("right")) ? -1 : 1;

        for (int i = start; i >= 0 && i < size; i += step) {
            for (int j = start; j >= 0 && j < size; j += step) {
                if (grid[i][j] == null) continue;

                int[] newPos = getNewPosition(i, j, direction);
                int ni = newPos[0], nj = newPos[1];

                while (ni >= 0 && ni < size && nj >= 0 && nj < size && grid[ni][nj] == null) {
                    grid[ni][nj] = grid[i][j];
                    grid[i][j] = null;
                    i = ni;
                    j = nj;
                    newPos = getNewPosition(i, j, direction);
                    ni = newPos[0];
                    nj = newPos[1];
                }

                if (ni >= 0 && ni < size && nj >= 0 && nj < size &&
                        grid[ni][nj] != null &&
                        grid[ni][nj].getValue() == grid[i][j].getValue() &&
                        !merged[ni][nj]) {
                    
                    grid[ni][nj].doubleValue();
                    grid[i][j] = null;
                    merged[ni][nj] = true;
                }
            }
        }

        spawnNewTile();
        return true;
    }

    public void printGrid() {
        for (Block[] row : grid) {
            for (Block block : row) {
                System.out.print((block == null ? 0 : block.getValue()) + "\t");
            }
            System.out.println();
        }
        System.out.println();
    }
}


// ==================== GAME CONTROLLER ====================
class GameController {
    private final Grid grid;
    private final Map<String, MoveStrategy> moveStrategies;

    public GameController() {
        this.grid = new Grid(4);
        this.moveStrategies = new HashMap<>();
        moveStrategies.put("up", new MoveUp());
        moveStrategies.put("down", new MoveDown());
        moveStrategies.put("left", new MoveLeft());
        moveStrategies.put("right", new MoveRight());
    }

    public void playMove(String direction) {
        if (!moveStrategies.containsKey(direction)) {
            System.out.println("Invalid move");
            return;
        }

        if (moveStrategies.get(direction).move(grid)) {
            grid.printGrid();
        } else {
            System.out.println("Move not possible!");
        }
    }
}


// ==================== MAIN GAME LOOP ====================
public class Game2048 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        GameController game = new GameController();

        while (true) {
            System.out.print("Enter move (up/down/left/right): ");
            String move = scanner.nextLine().trim().toLowerCase();
            game.playMove(move);
        }
    }
}
